### PR TITLE
[FLINK-9679] Implement AvroSerializationSchema

### PR DESCRIPTION
**Pull request by medcv**
## What is the purpose of the change

Provides implementation of AvroSerializationSchema that write records serialized as avro and also provides version that uses Confluent Schema Registry to write the record.

This is following AvroDESerializationSchema implementation patterns to have a consistent code base for Ser/Des.

## Brief change log

- Implemented AvroSerializationSchema / RegistryAvroSerializationSchema / ConfluentRegistryAvroSerializationSchema

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (yes)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by medcv** [comment]
@tillrohrmann I implemented `ConfluentRegistryAvroSerializationSchema` in this PR. I would appreciate if you review.

**Comment by dawidwys** [comment]
Hi @medcv ,
First of all thanks for the contribution. There is a couple of problems with this implementation though:
* ConfluentSchemaRegistry follows the idea that schemas are grouped by topic name. Providing the topic name in ctor of `SerializationSchema` will break this assumption when used with topic pattern. I think we should somehow enable passing schema from `KafkaProducer` to `SerializationSchema`, but it would require far more changes.
* You assume in `RegistryAvroSerializationSchema` that schema is always represented with a single  id(`int`). That is true only for the ConfluentSchemaRegistry, e.g. HortonworksSchemaRegistry uses metadata-id(`long`) and version id(`int`). The idea of `SchemaCoder` was to abstract away all "vendor" specific features.

**Comment by medcv** [comment]
@dawidwys Thanks a lot for the review.
I will start look into both problems that you've pointed out.
Second point would be easy to fix but for the first one I need do some research and see the feasibility of it.

**Comment by medcv** [comment]
@dawidwys I update the PR, please review

the usage would be like this
` ConfluentRegistryAvroSerializationSchema.forSpecific(User.class, subject,  schemaRegistryUrl)`
as Confluent needs "subject"  to fetch the Schema info. Now `ConfluentRegistryAvroSerializationSchema` uses "subject" directly instated on `topic + "-value"`.

**Comment by dawidwys** [comment]
@medcv the new commit does not address any of my previous comments or I don't understand something.

**Comment by medcv** [comment]
@dawidwys For second issue I am looking at other Schema registries and trying to extend `SchemaCoder`

**Comment by medcv** [comment]
@dawidwys 
in last commit, I did extend `SchemaCoder` to have `getSchemaId` as you suggested.

**Comment by dawidwys** [comment]
Actually I've suggested to move the whole writing schema logic into the coder. So the coder should have a method like `writeSchema(..., OutputStream out)`. 

Substituting previous `topic + "value"` with `subject` does not change anything. It simply renames a lookup key. The whole problem is that it is static and does not take into account the actual topic the message will be written to. Unfortunately I am afraid it would require some changes in the `FlinkKafkaProducer` itself.

**Comment by medcv** [comment]
@dawidwys Thanks for your suggestions.
New commit extend `SchemaCoder` with `writeSchema` method that helps to move the writing schema logic away from `AvroSerializationSchema` as you suggested.  

Totally agree with you that having dynamic `subject` variables make the implementations more generic but as this is `Confluent` specific implementations and this variable is only presented for `ConfluentRegistryAvroSerializationSchema`, I think a user of this method should be aware of how `Confluent` requires this variable when they setup their Kafka Producer and Schema Registry.  

I am open to suggestions to fix the issue ( by changing `FlinkKafkaProducer`) if you still thinking this is a blocker for this PR.

**Comment by medcv** [comment]
@tillrohrmann Please review

**Comment by medcv** [comment]
@dawidwys I would appreciate if you take a look for the latest changes on this PR.

New commit extend `SchemaCoder` with `writeSchema` method that helps to move the writing schema logic away from `AvroSerializationSchema` as you suggested.

Still using static `subject` on `ConfluentRegistryAvroSerializationSchema`

**Comment by tillrohrmann** [comment]
@medcv Dawid is currently on vacation and will be back next week. He will give it another pass when he's back. Please bear with us until then.

**Comment by medcv** [comment]
@tillrohrmann Thanks for the update!

**Comment by medcv** [comment]
@tillrohrmann I would appreciate if you take a look also. In last commit all the concerns has been address regarding using `SchemaCoder`. 

Only remaining concern is passing `subject` to the `ConfluentRegistryAvroSerializationSchema` which related to how Confluent Schema Registry works. As this is only specific to `ConfluentRegistryAvroSerializationSchema` I believe it should be ok to pass `subject` to the ctor.

**Comment by medcv** [comment]
@dawidwys Thanks for the review. I will do my best to address your comments on next commit and will add tests when the implementations got approved.

File: flink-formats/flink-avro-confluent-registry/pom.xml [hunk 1]
```diff
@@ -29,6 +29,10 @@ under the License.
 
 	<artifactId>flink-avro-confluent-registry</artifactId>
 
+	<properties>
+		<confluent.schema.rigestry.version>4.1.0</confluent.schema.rigestry.version>
```
Comment (WheresAlice) [inline flink-formats/flink-avro-confluent-registry/pom.xml]: Typo - "registry" not "rigestry"

**Comment by Wosin** [comment]
@medcv  
Hey, any update on this?
I am currently working on a project where we want to use Avro with Confluent and this would be a great thing to have.
Best Regards.

**Comment by dawidwys** [comment]
Subsumed by: #8371

**Pull request closed without merging**